daemon/daemon.go

```diff
@@ -131,10 +131,17 @@ func (d *Daemon) RemoveDriver(language string) error {
 }
 
 func (d *Daemon) DriverPool(ctx context.Context, language string) (*DriverPool, error) {
+	d.mu.RLock()
+	dp, ok := d.pool[language]
+	d.mu.RUnlock()
+	if ok {
+		return dp, nil
+	}
+
 	d.mu.Lock()
 	defer d.mu.Unlock()
-
-	if dp, ok := d.pool[language]; ok {
+	dp, ok = d.pool[language]
```

Comment:
This is a little subtle, and probably deserves a comment: If you reach this point the original version of `dp` you read at line 135 is actually fine—but it's possible it has since been evicted (between releasing the reader lock and acquiring the writer lock). It isn't clear why you couldn't just _use_ `dp` at that point—but I assume the reason is that you want to make sure you never use a driver that isn't mapped?

Reply:
Actually no, I take it back: `dp` is _not_ fine at this point.

Reply:
Correct. Should I write the comment for it? I assumed it's more a less a common pattern.

Reply:
No need, it's fine. It's not that common—but it's true that upgrading a lock is certainly a reasonable thing to do and can't be done in one step.